File: frontend/src/auth/pkce.ts
export async function generatePkce(): Promise<{ verifier: string; challenge: string }> {
  const bytes = crypto.getRandomValues(new Uint8Array(32))
  const verifier = base64Url(bytes)
  const digest = await crypto.subtle.digest('SHA-256', new TextEncoder().encode(verifier))
  return { verifier, challenge: base64Url(new Uint8Array(digest)) }
}

function base64Url(data: Uint8Array): string {
  let binary = ''
  data.forEach((b) => {
    binary += String.fromCharCode(b)
  })
  return btoa(binary).replaceAll('+', '-').replaceAll('/', '_').replaceAll('=', '')
}

export type TokenSet = {
  accessToken: string
  idToken?: string
  refreshToken?: string
  expiresAt: number
}

export function parseTokenResponse(json: Record<string, unknown>): TokenSet {
  const expiresIn = Number(json.expires_in ?? 3600)
  return {
    accessToken: String(json.access_token ?? ''),
    idToken: json.id_token ? String(json.id_token) : undefined,
    refreshToken: json.refresh_token ? String(json.refresh_token) : undefined,
    expiresAt: Date.now() + expiresIn * 1000,
  }
}

export function authorizeUrl(input: {
  supabaseUrl: string
  redirectUri: string
  challenge: string
}): string {
  const url = new URL(`${input.supabaseUrl.replace(/\/$/, '')}/auth/v1/authorize`)
  url.searchParams.set('redirect_to', input.redirectUri)
  url.searchParams.set('code_challenge', input.challenge)
  url.searchParams.set('code_challenge_method', 'S256')
  return url.toString()
}

export async function exchangeCode(input: {
  supabaseUrl: string
  publishableKey: string
  code: string
  verifier: string
}): Promise<TokenSet> {
  return tokenRequest(input.supabaseUrl, input.publishableKey, {
    auth_code: input.code,
    code_verifier: input.verifier,
  }, 'pkce')
}

export async function refreshTokens(input: {
  supabaseUrl: string
  publishableKey: string
  refreshToken: string
}): Promise<TokenSet> {
  return tokenRequest(input.supabaseUrl, input.publishableKey, {
    refresh_token: input.refreshToken,
  }, 'refresh_token')
}

async function tokenRequest(
  supabaseUrl: string,
  publishableKey: string,
  body: Record<string, string>,
  grantType: 'pkce' | 'refresh_token',
): Promise<TokenSet> {
  const res = await fetch(`${supabaseUrl.replace(/\/$/, '')}/auth/v1/token?grant_type=${grantType}`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      apikey: publishableKey,
    },
    body: JSON.stringify(body),
  })
  const json = (await res.json()) as Record<string, unknown>
  if (!res.ok) {
    throw new Error(String(json.error_description ?? json.error ?? json.msg ?? 'token exchange failed'))
  }
  return parseTokenResponse(json)
}

File: mobile/lib/auth/pkce.dart
import 'dart:convert';
import 'dart:math';

import 'package:crypto/crypto.dart';

/// 認可コード + PKCE（design.md D20）の検証子と challenge。
class PkcePair {
  const PkcePair({required this.verifier, required this.challenge});

  final String verifier;
  final String challenge;
}

/// frontend/src/auth/pkce.ts と同じ手順で S256 の組を作る。
PkcePair generatePkce([Random? random]) {
  final source = random ?? Random.secure();
  final bytes = List<int>.generate(32, (_) => source.nextInt(256));
  final verifier = base64UrlEncodeNoPad(bytes);
  final challenge = base64UrlEncodeNoPad(sha256.convert(utf8.encode(verifier)).bytes);
  return PkcePair(verifier: verifier, challenge: challenge);
}

String base64UrlEncodeNoPad(List<int> bytes) =>
    base64Url.encode(bytes).replaceAll('=', '');

/// GoTrue のトークン応答。
class TokenSet {
  const TokenSet({
    required this.accessToken,
    required this.expiresAt,
    this.idToken,
    this.refreshToken,
  });

  factory TokenSet.fromTokenResponse(Map<String, Object?> json, {DateTime? now}) {
    final expiresIn = switch (json['expires_in']) {
      final int value => value,
      final String value => int.tryParse(value) ?? 3600,
      _ => 3600,
    };
    return TokenSet(
      accessToken: json['access_token'] as String? ?? '',
      idToken: json['id_token'] as String?,
      refreshToken: json['refresh_token'] as String?,
      expiresAt: (now ?? DateTime.now()).add(Duration(seconds: expiresIn)),
    );
  }

  factory TokenSet.fromJson(Map<String, Object?> json) => TokenSet(
        accessToken: json['accessToken'] as String? ?? '',
        idToken: json['idToken'] as String?,
        refreshToken: json['refreshToken'] as String?,
        expiresAt: DateTime.parse(json['expiresAt'] as String),
      );

  final String accessToken;
  final String? idToken;
  final String? refreshToken;
  final DateTime expiresAt;

  /// 失効の 60 秒前から更新の対象にする。
  bool isFresh({DateTime? now}) =>
      expiresAt.difference(now ?? DateTime.now()) > const Duration(seconds: 60);

  Map<String, Object?> toJson() => {
        'accessToken': accessToken,
        'idToken': idToken,
        'refreshToken': refreshToken,
        'expiresAt': expiresAt.toIso8601String(),
      };
}

String _trimBase(String supabaseUrl) => supabaseUrl.replaceAll(RegExp(r'/$'), '');

/// GoTrue の認可エンドポイント URL を組み立てる。
Uri authorizeUrl({
  required String supabaseUrl,
  required String redirectUri,
  required String challenge,
  String provider = 'github',
}) =>
    Uri.parse('${_trimBase(supabaseUrl)}/auth/v1/authorize').replace(
      queryParameters: {
        'provider': provider,
        'redirect_to': redirectUri,
        'code_challenge': challenge,
        'code_challenge_method': 'S256',
      },
    );
